### PR TITLE
stats: refine estimation for index between row count

### PR DESCRIPTION
**Pull request by lamxTyler**
When the values that fall into the same bucket are not continuous, the current estimation may lead to very high inaccuracy, especially when the first column of the index has very small NDV, and other columns have larger NDV. For the multi-column index, this PR fixes it by using the column stats to estimate when the fraction of values [a, b) that falls into the bucket is quite small.
PTAL @coocood @zz-jason @winoros @fipped

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -385,13 +385,18 @@ func (hg *Histogram) greaterAndEqRowCount(value types.Datum) float64 {
 
 // lessRowCount estimates the row count where the column less than value.
 func (hg *Histogram) lessRowCount(value types.Datum) float64 {
+	count, _, _ := hg.lessRowCountWithFraction(value)
+	return count
+}
+
+func (hg *Histogram) lessRowCountWithFraction(value types.Datum) (float64, int, float64) {
```
Comment (fipped) [inline statistics/histogram.go]: I think this function needs a comment to explain what will return.

File: statistics/table.go [hunk 2]
```diff
@@ -356,8 +356,60 @@ func (t *Table) GetRowCountByIndexRanges(sc *stmtctx.StatementContext, idxID int
 		return getPseudoRowCountByIndexRanges(sc, indexRanges, float64(t.Count), colsLen)
 	}
 	result, err := idx.getRowCount(sc, indexRanges)
-	result *= idx.getIncreaseFactor(t.Count)
-	return result, errors.Trace(err)
+	if result >= 0 || err != nil {
+		return result * idx.getIncreaseFactor(t.Count), errors.Trace(err)
+	}
+	// cannot use the index stats to estimate, fall back to use the column stats.
+	return t.getIndexRowCountByColumns(sc, idx.Info, indexRanges)
```
Comment (winoros) [inline statistics/table.go]: What if we have index stats but don't have column stats?

Comment (lamxTyler) [reply statistics/table.go]: In this case, both estimations are inaccurate, we cannot know which one is better...

File: statistics/table.go [hunk 3]
```diff
@@ -356,8 +356,60 @@ func (t *Table) GetRowCountByIndexRanges(sc *stmtctx.StatementContext, idxID int
...
+			}
+
+			ranges = append(ranges, &ranger.Range{LowVal: []types.Datum{low}, HighVal: []types.Datum{high},
+				LowExclude: lowExclude, HighExclude: highExclude})
+		}
+		count, err := t.GetRowCountByColumnRanges(sc, t.colIDByName(col.Name.L), ranges)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		selectivity = selectivity * count / float64(t.Count)
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply statistics/table.go]: No, because when plan gets the table stats, if it finds the `Count` is 0, it will just use pseudo stats.

**Comment by lamxTyler** [comment]
Close it because we have a better solution now.

**Pull request closed without merging**